### PR TITLE
add stormlib package + example

### PR DESCRIPTION
**Pull request by wheybags**
Can be moved to a hunter-packages fork

### CODE REVIEW

**Comment by ruslo** [comment]
Created: https://github.com/hunter-packages/StormLib

**Comment by wheybags** [comment]
moved

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1966
* https://travis-ci.org/ingenue/hunter/builds/254740745?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
> /home/travis/build/ingenue/hunter/examples/stormlib/main.cpp:1:22: fatal error: Stormlib.h: No such file or directory

**Comment by wheybags** [comment]
Should be fixed now. Their example code had the wrong case in the header filename. windows...

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1990
* https://travis-ci.org/ingenue/hunter/builds/256076288?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
```
Undefined symbols for architecture x86_64:
  "_ltc_mp", referenced from:
```

**Comment by wheybags** [comment]
Ok, finally got an osx vm up to debug this, and I don't get what's going on here at all. It works on every other platform, and I used an object inspector to make sure the symbol was there and it is.
nm -gU on the libtomcrypt static lib outputs:
```
/Users/wheybags/stormlib/build/hunter/_Base/xxxxxxx/f77a954/3b0b532/Install/lib/libtomcryptd.a(crypt_ltc_mp_descriptor.c.o):
0000000000000188 C _ltc_mp
```

It's just a struct, declared here: https://github.com/hunter-packages/libtomcrypt/blob/1.17-p1/src/misc/crypt/crypt_ltc_mp_descriptor.c

and declared as extern here: https://github.com/hunter-packages/libtomcrypt/blob/1.17-p1/src/headers/tomcrypt_math.h#L416

Is there some weirdness on osx around extern structs? Do I need to do some equivalent of the windows style declspec dllimport crap?

@ruslo any idea what's going on here, because I'm stumped :(

**Comment by ruslo** [comment]
Library source is C and you're using it in C++ example. I guess we need `extern C`:
* https://stackoverflow.com/q/3789340/2288008

**Comment by wheybags** [comment]
But the whole thing is only ever included through tomcrypt.h: https://github.com/hunter-packages/libtomcrypt/blob/1.17-hunter/src/headers/tomcrypt.h#L73 , which has an extern c wrapper: https://github.com/hunter-packages/libtomcrypt/blob/1.17-hunter/src/headers/tomcrypt.h#L15

**Comment by ruslo** [comment]
> But the whole thing is only ever included through tomcrypt.h

Are we talking about tomcrypt or stormlib? :)

In `examples/stormlib/main.cpp` I see `#include <StormLib.h>`.

**Comment by wheybags** [comment]
I'm not sure which package contains the problem :(
Stormlib links to libtomcrypt, and uses that symbol in only two places https://github.com/hunter-packages/StormLib/blob/v9.21_hunter/src/SBaseCommon.cpp#L187 and https://github.com/hunter-packages/StormLib/blob/v9.21_hunter/src/libtomcrypt/src/pk/rsa/rsa_verify_simple.c (disregard the weird path on the second one, it was using a bundled version of the lib with that file added in).

Those files are getting the definition of the ltc_mp symbol from https://github.com/hunter-packages/StormLib/blob/v9.21_hunter/src/StormCommon.h#L49 and https://github.com/hunter-packages/StormLib/blob/v9.21_hunter/src/libtomcrypt/src/pk/rsa/rsa_verify_simple.c#L11 , so it should be picking up on the extern "C" from there, and indeed it works on linux and windows.

**Comment by ruslo** [comment]
Looks weird indeed. Add usage of `ltc_mp` to `tomcrypt` example and you will see the same error so it's not `stormlib` issue. May be some mess with defines? While building `tomcrypt` we are using one set and while using `tomcrypt` another set? And linker think that those structs `ltc_math_descriptor` don't match?

**Comment by wheybags** [comment]
oooook, so (hesistant to say this but), this looks like it might a linker bug?

I added an extra symbol into that file, so:
```patch
diff --git a/src/headers/tomcrypt_math.h b/src/headers/tomcrypt_math.h
index aee6105..ba60fba 100644
--- a/src/headers/tomcrypt_math.h
+++ b/src/headers/tomcrypt_math.h
@@ -414,6 +414,7 @@ typedef struct {
 } ltc_math_descriptor;
 
 extern ltc_math_descriptor ltc_mp;
+extern int testInt;
 
 int ltc_init_multi(void **a, ...);
 void ltc_deinit_multi(void *a, ...);
diff --git a/src/misc/crypt/crypt_ltc_mp_descriptor.c b/src/misc/crypt/crypt_ltc_mp_descriptor.c
index 0577d1d..f0c781f 100644
--- a/src/misc/crypt/crypt_ltc_mp_descriptor.c
+++ b/src/misc/crypt/crypt_ltc_mp_descriptor.c
@@ -11,3 +11,5 @@
 #include "tomcrypt.h"
 
 ltc_math_descriptor ltc_mp;
+
+int testInt = 500;
```

^Added testInt to the .c and the header.

Then I made an example program:

```c
#include <tomcrypt.h>

int main(int argc, char** argv)
{
    ltc_mp = ltm_desc;  
    printf("%d\n", testInt);

    printf("%s\n", ltc_mp.name);
    return 0;
}
```

And this works just fine, links perfectly and prints: 
```
500
LibTomMath
```
As you might expect.

But, if I comment out the first printf, the one using testInt, it fails to link.
So it seems like the linker is ignoring that file unless there is another symbol in there that's being used besides ltc_mp.

So, yeah, how should I proceed on this? Am I misunderstanding some subtlety of the osx linker?

**Comment by ruslo** [comment]
> oooook, so (hesistant to say this but), this looks like it might a linker bug?

Might be. Try to create [SSCCE](http://www.sscce.org/) and if linker will fail in this case you can open bugreport.

> Am I misunderstanding some subtlety of the osx linker?

I don't think so. Code looks good, error is weird.

> So, yeah, how should I proceed on this?

Up to you. You can exclude OSX from testing or if adding dummy `extern int` will fix the issue you can add this as a workaround.

**Comment by wheybags** [comment]
Minimal example here: https://gist.github.com/wheybags/0cf9abe991d1c667e5218bf59ab24649
I'm using yosemite, if you have access to the newest version of osx/xcode, would you mind verifying this?

**Comment by ruslo** [comment]
> Minimal example here

Makefile, really? :)

> I'm using yosemite, if you have access to the newest version of osx/xcode, would you mind verifying this?

Latest Xcode 9 available on Travis.

GitHub + Travis:
* https://github.com/forexample/osx-extern-struct-linker-error

**Comment by wheybags** [comment]
Hey, it was either that or a shell script :p
Ok, I'll send a bug report to apple I guess, and in the meantime I'll add
some nasty hacky workarounds to the libtomcrypt and stormlib packages
Thanks!

**Comment by wheybags** [comment]
Ok, so updated with a workaround someone pointed out to me on IRC (forgot your name workaround guy, sorry :( )

File: cmake/configs/default.cmake [hunk 1]
```diff
@@ -108,11 +108,12 @@ hunter_config(SDL2 VERSION 2.0.4-p4)
 hunter_config(SDL_mixer VERSION 2.0.1-p1)
 hunter_config(SQLite3 VERSION autoconf-3080803) #R-Tree enabled
 hunter_config(Sober VERSION 0.1.3)
+hunter_config(stormlib VERSION 9.21-p1)
```
Comment (ruslo) [inline cmake/configs/default.cmake]: Please split these changes.

**Comment by wheybags** [comment]
done: https://github.com/ruslo/hunter/pull/971

**Comment by ruslo** [comment]
> done: #971

Merged, please rebase.

**Comment by wheybags** [comment]
done

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2202
* https://travis-ci.org/ingenue/hunter/builds/266315803?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
> Testing

Some toolchains failed, feel free to exclude them from testing.

**Comment by wheybags** [comment]
disabled analyze: https://github.com/ingenue/hunter/pull/113
Also, updated, should fix MSYS2 build

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2214
* https://travis-ci.org/ingenue/hunter/builds/266526369?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
> should fix MSYS2 build

Still failed:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2214/job/y6pw3biw4eymeesl

**Comment by wheybags** [comment]
Ok, updated again, should work now.

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2239
* https://travis-ci.org/ingenue/hunter/builds/268638207?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
Uploading:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2241
* https://travis-ci.org/ingenue/hunter/builds/268642116?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
Released: https://github.com/ruslo/hunter/releases/tag/v0.19.83

**Pull request merged**